README.md
# ino2cpp
[![Build Status](https://github.com/tj/commander.js/workflows/build/badge.svg)](https://github.com/tj/commander.js/actions?query=workflow%3A%22build%22)<a href="https://github.com/MavrickUK/ino2cpp/tags" rel="nofollow">
    <img alt="GitHub tag (latest SemVer pre-release)" src="https://img.shields.io/github/v/tag/MavrickUK/ino2cpp?include_prereleases&label=version"/>
  </a><a href="https://github.com/MavrickUK/ino2cpp/issues">
    <img src="https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat" alt="CodeFactor" />
  </a>

Convert Arduino INO sketches to C++ by creating new .cpp and .h files


## Details

Arduino sketches and C++ are very similar.
However, an INO file cannot be compiled as-is by C/C++ compilers (e.g. GCC).
This tool converts INO sketches to C++ code such that off-the-shelf compilers and static analysis tools can be executed on the code.

## Converter
There are three steps in this conversion [1][2]:****
1. **Generate forward declarations**. Arduino INO sketches allow the use of a function before its definition. The first step is to parse the INO sketch to obtain function signatures, and generate a header file with these signatures ("sketch_name.h").
2. **Includes**. Two includes are inserted before the content of the INO sketch: #include <Arduino.h>, and #include "sketch_name.h".
3. **Write C++ to disk**. Write the resulting C++ and header file to disk. INO is unchanged!

## Installation
Portable, just download the latest release version and run the .exe from where you unzipped.
**Or** you can use **Scoop** package manager and install with:
1. Install [scoop](https://github.com/lukesampson/scoop)
2. Add this bucket to scoop:
```bash
scoop bucket add ino2cpp https://github.com/MavrickUK/ino2cpp
```
3. Install tools via `scoop install`:
```bash
scoop install ino2cpp
```
4. You're done.
## Usage
```
ino2cpp <filename> [-o <filename>] [-v]
ino2cpp uno_led.ino
ino2cpp -h
```

## Example
```
➜ .\ino2cpp.exe uno_led.ino -v
Ino2Cpp Converter v0.2
Working, please wait...
  void update_started();
  void update_finished();
  void update_progress(int cur, int total);
  void update_error(int err);
  void otaUpdate(String url);
  void check_for_updates();
Funcs exported: 6
Added: #include <Arduino.h> and #include "uno_led.h" to uno_led.cpp
uno_led.cpp and uno_led.h created. Done!
```

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

## References
[1] [How to convert Arduino example sketch to a complete c++ project?](https://arduino.stackexchange.com/questions/32998/how-to-convert-arduino-example-sketch-to-a-complete-c-project)
[2] [converting .ino files to .cpp - Arduino Forum](https://forum.arduino.cc/t/converting-ino-files-to-cpp/226366)

### License
[MIT](https://choosealicense.com/licenses/mit/)
****


ChangeLog.md

# Change Log
All notable changes to this project will be documented in this file.
If you find a bug, please open and issue in the github repo.

## [0.3] - 29-03-2023

### Added
- builds for 32/64bit.

### Fixed
- Minor issues.
- Some typos.


## [0.2] - 21-03-2023

### Added
- Validate input filename for invalid characters.
- Included some unit testing for some key function.
- New info flag [-i].
- About command.
- Filenames checked for weird characters and replaced.

### Changed
- The [-i] flag now displays process info.
- Updated ReadMe.md
- Input filename with or without suffix.

### Fixed
- Minimum arguments. Inform user.
- Double use of [-v] flag. Changed one into [-i].


## [0.1] - 15-03-2023

Initial release of ino2cpp.

utils/validFilename_test.go
package utils

import "testing"

func TestRemoveInvalidFilenameChars(t *testing.T) {
	type args struct {
		fn string
	}
	tests := []struct {
		name string
		args args
		want string
	}{
		{
			name: "Test case 1",
			args: args{fn: "file?name*with<>invalid|chars"},
			want: "file_name_with_invalid_chars",
		},
		{
			name: "Test case 2",
			args: args{fn: "filename_with_no_invalid_chars"},
			want: "filename_with_no_invalid_chars",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := RemoveInvalidFilenameChars(tt.args.fn); got != tt.want {
				t.Errorf("RemoveInvalidFilenameChars() = %v, want %v", got, tt.want)
			}
		})
	}
}


parser/parseINO.go
package parser

import (
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

//goland:noinspection ALL
const (
	cARDUINOH = "#include <Arduino.h>"
	cREGEX    = `(void \w+\(\s*(?:\w+\s+\w+\s*(?:,\s*)?)*\s*\))\s*\{`
)

var lines []string

type Parse struct {
	sourceFilename string
	outputFilename string
	verboseOutput  bool
}

func NewParse(fname string, oname string, verbose bool) *Parse {
	p := &Parse{
		sourceFilename: strings.TrimSuffix(filepath.Base(fname), filepath.Ext(fname)),
		outputFilename: oname,
		verboseOutput:  verbose,
	}
	return p
}

func (p *Parse) Start(appVersion string) {
	fmt.Printf("Ino2Cpp Converter v%s\n", appVersion)
	fmt.Println("Working, please wait...")

	contentChan := make(chan []byte)
	errChan := make(chan error)

	// Read the file asynchronously
	go func() {
		content, err := os.ReadFile(p.sourceFilename + ".ino")
		if err != nil {
			errChan <- fmt.Errorf("error reading file: %w", err)
			return
		}
		contentChan <- content
	}()

	// Wait for the file to be read
	select {
	case content := <-contentChan:
		matchFunctions(content, p.verboseOutput)

		p.createHeader(p.outputFilename)
		p.createCppFile(p.sourceFilename)

		fmt.Printf("%s.cpp and %s.h created. Done!\n", p.outputFilename, p.outputFilename)

	case err := <-errChan:
		fmt.Println(err)
	}
}

func matchFunctions(content []byte, verbose bool) {
	input := string(content)
	pattern := cREGEX
	var funcsFound int
	r := regexp.MustCompile(pattern)
	matches := r.FindAllStringSubmatch(input, -1)
	for _, match := range matches {
		if len(match) > 1 {
			if !((strings.Contains(match[1], "setup")) ||
				(strings.Contains(match[1], "loop"))) {
				if verbose {
					fmt.Println("  " + match[1] + ";")
				}
				lines = append(lines, match[1]+";")
				funcsFound++
			}
		}
	}
	fmt.Printf("Funcs exported: %d\n", funcsFound)
}

// Creates the .h file containing all our exported functions
func (p *Parse) createHeader(fn string) {
	// Create a file for writing
	f, err := os.Create(fn + ".h")
	if err != nil {
		log.Fatalf("error creating file: %v", err)
	}
	defer f.Close()

	// Use a strings.Builder instead of a bufio.Writer
	var b strings.Builder

	for i := range lines {
		_, err := b.WriteString(lines[i] + "\n")
		if err != nil {
			log.Fatalf("error writing to file: %v", err)
		}
	}

	// Write the contents of the builder to the file
	_, err = io.WriteString(f, b.String())
	if err != nil {
		log.Fatalf("error writing to file: %v", err)
	}
}

// Modify the original file and add two lines to the start
func (p *Parse) createCppFile(fn string) {
	inputFile, err := os.OpenFile(fn+".ino", os.O_RDONLY, 0644)
	if err != nil {
		panic(err)
	}
	defer inputFile.Close()

	outputFile, err := os.Create(p.outputFilename + ".cpp")
	if err != nil {
		panic(err)
	}
	defer outputFile.Close()

	// Write the additional content at the beginning of the output file
	header := []byte(cARDUINOH + "\n" + `#include "` + p.outputFilename + ".h" + `"` + "\n\n")
	_, err = outputFile.Write(header)
	if err != nil {
		panic(err)
	}

	// Copy the contents of the input file to the output file
	_, err = io.Copy(outputFile, inputFile)
	if err != nil {
		panic(err)
	}

	if p.verboseOutput {
		fmt.Printf("Added: %s and #include \"%s.h\" to %s.cpp\n", cARDUINOH, p.outputFilename, p.outputFilename)
	}
}


cmd/rootCmd.go
package cmd

import (
	"fmt"
	"github.com/spf13/cobra"
	"ino2cpp/parser"
	"ino2cpp/utils"
	"os"
	"strings"
)

const (
	AppName         = "ino2cpp"
	AppVersion      = "0.3" //TODO: Update BEFORE release/push
	BuildDate       = "29 Mar 2023"
	cFilenameSuffix = ".ino"
	GitHubRepo      = "https://github.com/MavrickUK/ino2cpp"
)

var (
	//inoFilename string
	outFilename string
	verbose     bool
)

// rootCmd represents the base command when called without any subcommands
var (
	rootCmd = &cobra.Command{
		Use:   "ino2cpp <filename> [-o <filename>] [-i]",
		Short: "Convert Arduino INO sketches to C++",
		Long: `Arduino sketches and C++ are very similar.
However, an INO file cannot be compiled as-is by C/C++ compilers (e.g. GCC).
This tool converts INO sketches to C++ code such that off-the-shelf compilers and static analysis tools can be executed on the code.
`,
		Example: `
  ino2cpp example.ino
  ino2cpp example.ino -o new_filename
  ino2cpp example.ino -i`,
		Args: cobra.MinimumNArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			startParsing(args[0]) // This should be the provided ino filename
		},
		Version: AppVersion,
	}
)

func startParsing(fn string) {
	fn = utils.RemoveInvalidFilenameChars(fn)
	fn = strings.TrimSuffix(fn, cFilenameSuffix)

	if outFilename == "" {
		outFilename = fn
	}

	p := parser.NewParse(fn, outFilename, verbose)
	p.Start(AppVersion)
}

// Execute adds all child commands to the root command and sets flags appropriately.
// This is called by main.main(). It only needs to happen once to the rootCmd.
func Execute() {
	if err := rootCmd.Execute(); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}

func init() {
	rootCmd.PersistentFlags().StringVarP(&outFilename, "output", "o", "", "output filename.")
	rootCmd.PersistentFlags().BoolVarP(&verbose, "info", "i", false, "result info")
}
